PXC-3312: Prevent cleanup of statement diagnostic area in case of tra…

kamil-holubicki opened the pull request:
…nsaction replay.

Fixed issues:
1. When transaction was marked to be replied its diagnostic area was cleaned up. Then, after replaying this cleaned diagnostic are was restored causing diagnostics to be empty.
2. MW-329 test stabilized. It was observed under heavy CPU load that test failed with timeout. It was caused be test flow, which expects at least 5 updates to end up with replay. However under high load it happened that test table grown to large for update transaction to be BF aborted after replication but before certification which ended up with deadlock error without replay. Modified test flow to prevent test table size to grow too much.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/MW-329.test:
```diff
@@ -1,15 +1,26 @@
 # #MW-329 Fix incorrect affected rows count after replay
 #
+# The following test starts insert thread on node_1 and parallel update thread on node_2
+# Then it is expected to update to be BF abborted on node_2 after it is replicated, but before it is certified.
```
1. Did you mean that 
```
# Then it is expected that update to be BF aborted on node_2 after it is replicated, but before it is certified.
```

2. `s/abborted/aborted/`

kamil-holubicki replied:
done

venkatesh-prasad-v commented on mysql-test/suite/galera/t/MW-329.test:
```diff
@@ -47,17 +91,26 @@ select sleep(0.5);
-			--die ROW_COUNT() = 0
-		}
-	}
+        # If we updated block was sig_ted, stop and allow deleteion from t1. 
```
`s/deleteion/deletion/`

kamil-holubicki replied:
done

venkatesh-prasad-v commented on mysql-test/suite/galera/t/MW-329.test:
```diff
@@ -1,15 +1,26 @@
+# Then it is expected to update to be BF abborted on node_2 after it is replicated, but before it is certified.
+# This will cause update transaction to be replayed.
+# However under high CPU load, when running this test like 
+# ./mtr MW-329{,,,,,,,,,,} --parallel=11 --mem --testcase-timeout=10 
```
Applying the patch shows these warnings. Please remove the trailing whitespaces in the end of these two lines.

```
1215.diff:67: trailing whitespace.
# However under high CPU load, when running this test like 
1215.diff:68: trailing whitespace.
# ./mtr MW-329{,,,,,,,,,,} --parallel=11 --mem --testcase-timeout=10 
1215.diff:147: trailing whitespace.
        # If we updated block was sig_ted, stop and allow deleteion from t1. 
warning: 3 lines add whitespace errors.
```

kamil-holubicki replied:
done

venkatesh-prasad-v commented on mysql-test/suite/galera/t/MW-329.test:
```diff
@@ -47,17 +91,26 @@ select sleep(0.5);
-			--die ROW_COUNT() = 0
-		}
-	}
+        # If we updated block was sig_ted, stop and allow deleteion from t1. 
```
Remove the trailing whitespace in the end.

kamil-holubicki replied:
done

venkatesh-prasad-v commented on sql/binlog.cc:
```diff
@@ -8452,7 +8452,7 @@ TC_LOG::enum_result MYSQL_BIN_LOG::commit(THD *thd, bool all) {
       thd_get_cache_mngr(thd)->reset();
       // Reset the thread OK status before changing the outcome.
-      if (thd->get_stmt_da()->is_ok())
+      if (!wsrep_must_replay(thd) && thd->get_stmt_da()->is_ok())
```
1. At this point, if thd already has any error set, then calling `my_error()` will make the DA will make the assert `!set` in `DA::set_error_status()`. Is it expected that there shall be no error when we reach here?

2. This will cause the thread to not reset the DA if the transaction must be replayed. So, if we are replaying the transaction, then from where does the DA get cleared before the replay?

kamil-holubicki replied:
1. The callstack here is
```
MYSQL_BIN_LOG::commit(THD*, bool)
ha_commit_trans(THD*, bool, bool)
trans_commit_stmt(THD*, bool)
mysql_execute_command(THD*, bool)
mysql_parse(THD*, Parser_state*, bool)
```

DA allows overwriting the result (I've put comment in details in the code)
 
2. Please take a looka at Wsrep_replayer_service constructor and destructor and m_da_shadow in wsrep_high_priority_service.cc

3. Commit message first line is Jira issue summary

venkatesh-prasad-v replied:
Regarding the comments on the commit message, I actually meant to wrap the message body width to 80 characters. something like below.

```
PXC-3312: Prevent cleanup of statement diagnostic area in case of transaction replay.

Fixed issues:
1. When transaction was marked to be replied its diagnostic area was cleaned up. 
Then, after replaying this cleaned diagnostic are was restored causing
diagnostics to be empty. 

2. MW-329 test stabilized. It was observed under heavy CPU load that test failed
with timeout. It was caused be test flow, which expects at least 5 updates to
end up with replay. However under high load it happened that test table grown to
large for update transaction to be BF aborted after replication but before
certification which ended up with deadlock error without replay. Modified test
flow to prevent test table size to grow too much.
```

kamil-holubicki replied:
done

venkatesh-prasad-v reviewed (commented):
Hi Kamil! 

Thank you for working on this issue. It is indeed a tricky one to fix. Nice job!

If you don't mind, could you please limit the commit message to 80 characters wide, just like other commit messages, so that is more readable?

Apart from that, I also have a few questions and a few suggestions. Please see below.

Thank you for your time.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/MW-329.test:
```diff
@@ -22,10 +33,43 @@ SELECT VARIABLE_VALUE = 0 FROM performance_schema.global_status WHERE VARIABLE_N
+                                SET @foo = (SELECT SLEEP(0.1));
+                                SELECT can_delete FROM sig_t INTO delete_allowed;
+                        END WHILE;
+		END IF;
```
If it isn't too much to ask, then please replace tabspaces with whitespaces for better readability.

venkatesh-prasad-v reviewed (approved):
Hi Kamil,

Thank you for addressing the comments. 
The patch looks good now and have a few cosmetic suggestions below.

Apart from that, it is approved from my side.

Thank you.

kamil-holubicki committed:
PXC-3312: Prevent cleanup of statement diagnostic area in case
of transaction replay.

Fixed issues:

1. When transaction was marked to be replayed its diagnostic area was
cleaned up. Then, after replaying this cleaned diagnostic are was
restored causing diagnostics to be empty and 0 rows affected reported
to the client.

2. MW-329 test stabilized. It was observed under heavy CPU load that
test failed with timeout. It was caused be test flow, which expects
at least 5 updates to end up with replay. However under high load it
happened that test table grown to large for update transaction to be
BF aborted after replication but before certification which ended up
with deadlock error without replay. Modified test flow to prevent test
table size to grow too much.
